Get-Command: Specifying an alias and -Syntax returns the aliased commands syntax

ChrisLGardner opened the pull request:
# PR Summary

Add an additional check into the `IsCommandMatch` method to return the syntax of an aliased command rather than just the name of the aliased command.

Previous behaviour:
![image](https://user-images.githubusercontent.com/16837142/66720667-d7847d00-edf7-11e9-944a-cae5e83f0d55.png)

New behaviour:
![image](https://user-images.githubusercontent.com/16837142/66720653-b4f26400-edf7-11e9-889a-c2624305f4df.png)


## PR Context

#5963 asked for this functionality as currently `Get-Command del -syntax` doesn't give you anything useful, especially not beyond what `Get-Command del` does. The Issue does suggest that instead of the original command it shows the alias in the syntax but the `Syntax` property is read-only and I didn't want to mess around with the base CommandInfo or AliasInfo object for something like this, not doing that also somewhat enables @KirkMunro's request that it shows the `alias -> real command` as part of the output since it'll now show the resolved command name in the syntax'

I don't think this is a breaking change because the likelihood of anyone actually using `Get-Command -Name <alias> -Syntax` for anything outside the command line is very minimal. I'm also not sure if there needs to be a docs change for it but it might be useful just so users are aware that it'll do this now.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

ChrisLGardner committed:
Add additional check for alias and syntax switch

Return aliased command syntax instead of the name of the aliased command

ChrisLGardner commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -1227,7 +1227,7 @@ private bool IsCommandMatch(ref CommandInfo current, out bool isDuplicate)
                 if (isCommandMatch)
                 {
-                    if (ArgumentList != null)
+                    if ((current.CommandType == CommandTypes.Alias) && Syntax)
```
General question for more knowledgeable people, would it be faster this way or  with the conditions flipped (`(Syntax && (current.CommandType == CommandTypes.Alias))`)? 

This'll run on every Get-Command call (that finds a match) so I feel like having them the other way around would be better since the number of times Syntax is specified is going to be much smaller so it'll fail fast through the condition. If someone can point me to a good guide on how to actually compare them then that'd be useful too. I'd guess the difference is minimal but if one is faster than the other then it makes more sense to use whichever it is.

vexx32 replied:
If the condition is `&&` there isn't really a "positive" short-circuit since both conditions have to be satisfied, but there may be good reason to negatively short-circuit the code path. If there is a condition which is more commonly `false` than the other one, that should be first.

vexx32 replied:
Also, I think it's probably a good idea to use `Syntax.IsPresent` so the compiler doesn't need to bother invoking the implicit cast to boolean behind the scenes.

iSazonov replied:
For me `Syntax.IsPresent && (current.CommandType == CommandTypes.Alias` is more readable.

ChrisLGardner replied:
Fixed.

ChrisLGardner replied:
That's a much cleaner solution. I've added that now.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:
```diff
@@ -103,3 +103,29 @@ Describe "Get-Command Feature tests" -Tag Feature {
 }
+
+Describe "Get-Command" -Tag CI {
+    Context "-Syntax tests" {
```
To make the tests reliable we need to load Microsoft.PowerShell.Management module.
Please add this in `BeforeAll` block

ChrisLGardner replied:
I think I've added it in the right place, I couldn't see any other examples of how to do it so figured a simple `Import-Module Microsoft.PowerShell.Management` would be enough.

iSazonov commented:
@daxian-dbw It seems there is a problem with TestAppDomainProcessExitEvenHandlerNotLeaking test. Could you please look?
https://powershell.visualstudio.com/PowerShell/_build/results?buildId=35043&view=logs&jobId=3057ddcd-9849-50b7-e99e-f5fc6d303bb4&taskId=ff5d329a-0734-5136-788e-29ee918190af&lineStart=9&lineEnd=24&colStart=1&colEnd=27

```c#
Description: PSTests.Sequential.RunspaceTests PSTests.Sequential.RunspaceTests PSTests.Sequential.RunspaceTests PSTests.Sequential.RunspaceTests
Name:        TestAppDomainProcessExitEvenHandlerNotLeaking TestRunspaceWithPowerShellAndInitialSessionState TestRunspaceWithPipeline TestRunspaceWithPowerShell
message:
Test-XUnitTestResults : Cannot validate argument on parameter 'message'. The argument is null or empty. Provide an 
argument that is not null or empty, and then try the command again.
At D:\a\_temp\85b7acd9-47dd-489a-a2cb-ec41b0000555.ps1:5 char:1
+ Test-XUnitTestResults -TestResultsFile $xUnitTestResultsFile
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Test-XUnitTestResults], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Test-XUnitTestResults
```

KirkMunro commented:
> The Issue does suggest that instead of the original command it shows the alias in the syntax but the `Syntax` property is read-only and I didn't want to mess around with the base CommandInfo or AliasInfo object for something like this, not doing that also somewhat enables @KirkMunro's request that it shows the `alias -> real command` as part of the output since it'll now show the resolved command name in the syntax'.

You don't need to change the `Syntax` property for this. `Get-Command ... -Syntax` returns a single string. All you need to do to show a more meaningful (and technically more correct) output is to modify the syntax strings you retrieve from the command associated with that alias, by replacing the leading command name with the name of the alias.

i.e. When you discover that it is an alias, store the alias name, and then later when you retrieve the command syntax, replace the command name with the alias name in the output, while also pre-pending that output with a comment showing the alias to command relationship.

This will be very helpful when an alias references a script path, because you'll see the alias in the output syntax (which is the command you requested syntax for), while also seeing that the command you're requesting syntax for is actually an alias for another command (good for awareness/educational purposes).

I'd really like to see the output from this match something more along these lines:

```powershell
# First, create a script file
@'
[CmdletBinding()]
param(
    [Parameter(Position=0, Mandatory, ValueFromPipeline, ValueFromPipelineByPropertyName)]
    [ValidateNotNullOrEmpty()]
    [string[]]
    $Name
)
process {
    "Processing ${Name}"
}
'@ | out-File ${env:USERPROFILE}\Test-GcmSyntax.ps1

# Now set up an alias for that file
new-alias tgs ${env:USERPROFILE}\Test-GcmSyntax.ps1

# Get the command syntax using the alias
gcm -syntax tgs
```

Desired output:

```none
Note: tgs is an alias for C:\Users\kirka\Test-GcmSyntax.ps1

tgs [-Name] <string[]> [<CommonParameters>]

```

To make that work, you can inject the alias definition at the top, and then use regex to replace `"^Test-GcmSyntax.ps1"` (or an actual command name if it was an alias for a command instead of for a script) with `"tgs"`.

If we don't agree on showing the alias definition at the top, at a minimum I would expect the syntax to show the alias name instead of the command/script it points to.

As an alternative, we could show the alias details in verbose output instead, so that users only see it when invoking the command with `-Verbose`. I don't love that though because people should realize when they are looking at an alias, so we really need to show them that detail. They see that when they invoke `Get-Command` without `-Syntax`, but they should really see it when they invoke it with `-Syntax` as well.

ChrisLGardner commented:
I've been digging into doing that a bit and I've managed to get this so far:

```
[547.63 ms]github\powershell [get-command-alias ≡ +0 ~1 -0 !]>  gcm del -syn
Note: del is an alias for Remove-Item

del [-Path] <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] 
[-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]

del -LiteralPath <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]
```

It doesn't bring back the full path for a script but I'll see what I can do about that.  I do like the extra information that shows what the alias is for in a more clear way than just replacing the name in the syntax output.

ChrisLGardner commented:
After a bit more fiddling around I've got it working with external scripts, I need to add some test cases for this and need to play around with it and native commands to see how that comes out. 
```
PS>  gcm -syntax tgs
Note: tgs is an alias for C:\Users\Chris\Test-GcmSyntax.ps1
tgs [-Name] <string[]> [<CommonParameters>]
```

vexx32 commented:
You can probably relay the information that it's an alias a bit more succinctly, similarly to how `Get-Alias` shows it; `tgs -> C:\Users\Chris\Test-GcmSyntax.ps1` or `gci -> Get-ChildItem`

I do think having a line break after the "hey this is an alias" line is a good idea too, it helps clarify that the information being given is in two parts. 🙂

ChrisLGardner commented:
> I do think having a line break after the "hey this is an alias" line is a good idea too, it helps clarify that the information being given is in two parts. 🙂

That's what I'm thinking, it'll take a bit more conditional formatting as the syntax for a command already includes the leading line breaks but I can add that in easily enough.

ChrisLGardner committed:
Add support for aliases to script files returning the correct syntax and location

KirkMunro commented:
> You can probably relay the information that it's an alias a bit more succinctly, similarly to how `Get-Alias` shows it; `tgs -> C:\Users\Chris\Test-GcmSyntax.ps1` or `gci -> Get-ChildItem`

Except with `Get-Alias`, you know what you're getting, so seeing the shorthand such as `gci -> Get-ChildItem` makes more sense. You have context in that scenario to work with the information that is presented to you.

With `Get-Command`, you don't have the type of the command in the output, so `gci -> Get-ChildItem` by itself may not make as much sense unless you're familiar with that output already.

KirkMunro commented:
Something shorter would be good though, especially if it doesn't need to be localized. Maybe something like this:

`gci (alias) -> Get-ChildItem`

`tgs (alias) -> C:\Users\Chris\Test-GcmSyntax.ps1`

You wouldn't need to quote the right-hand side of the `->` if the path contained spaces with that syntax.

iSazonov commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -531,6 +531,17 @@ private void OutputResultsHelper(IEnumerable<CommandInfo> results)
+                                    this.Name[0],
+                                    result is ExternalScriptInfo extScript ? string.Format("{0}{1}", extScript.Path, Environment.NewLine) : result.Name,
+                                    Environment.NewLine,
+                                    result.Syntax.Replace(result.Name, this.Name[0]));
```
And how does output look with the replace?

KirkMunro replied:
Also I see you checking for wildcards. Is the logic smart enough to show the alias properly if a wildcard search finds commands that are aliases and outputs the syntax for them? There should probably be a specific test just for that scenario to make sure it is covered.

ChrisLGardner replied:
For scripts it looks like this:

```
PS> gcm tgs -syn
tgs -> C:\Users\Chris\Test-GcmSyntax.ps1

tgs [-Name] <string[]> [<CommonParameters>]
```

For commands it's:

```
PS>  gcm del -syn
del -> Remove-Item

del [-Path] <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] 
[-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]

del -LiteralPath <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]
```

ChrisLGardner replied:
> Also I see you checking for wildcards. Is the logic smart enough to show the alias properly if a wildcard search finds commands that are aliases and outputs the syntax for them? There should probably be a specific test just for that scenario to make sure it is covered.

Not right now, I need to add something in for that.

With the codefactor tests complaining about it being too complex I'll break out this into it's own method as it'll need to handle this case too and that'll make it more difficult to follow I think.

iSazonov replied:
I don't like last commits. I think it is a task for Get-Help to return enhanced information in human format.

KirkMunro replied:
@iSazonov Actually, it's a task for `Get-Command`, and that's exactly what it does. `Get-Command gci` shows you that `gci` is an alias for `Get-Command`. With this PR, `Get-Command gci -Syntax` shows you the syntax for the `gci` command and also identifies that it's an alias for `Get-ChildItem`. There's no reason to force users to have to call `Get-Command` twice to get such important information.

In fact, it would probably be far more useful for a less experienced user if `-Syntax` reported this additional information, and if the user had `$PSDefaultParameterValues['Get-Command:Syntax'] = $true` in their profile, because unless they are programmatically working with command metadata or trying to find out what module a command comes from, this information is far more useful for them.

iSazonov replied:
> There's no reason to force users to have to call Get-Command twice to get such important information.

Get-Help is for users to get all in once. Get-Command is more for scripting - without -Syntax it returns typed object (AliasInfo or CommandInfo) and I'd even expect that -Syntax returns typed output too as a syntax tree.

KirkMunro commented:
In theory, maybe, but in practice `Get-Command` is also taught as a learning tool, and it is used interactively with `-Syntax` very, very frequently as a way to simply get the syntax of a command. It also just returns a `string`.

That said, this part of the discussion makes me wonder if `Get-Help` should have a `-Syntax` parameter. The challenge there is muscle memory for users who have been using `gcm -syntax` for quick command inspection for years...

kilasuit commented:
> that said, this part of the discussion makes me wonder if Get-Help should have a -Syntax parameter. 

I think maybe but thats a separate issue than this and whilst get-command already has syntax  (and is in muscle memory) let's get improvements made to it

ChrisLGardner committed:
Refactor output creation to its own method to reduce complexity within the helper method

ChrisLGardner commented:
I've added support for wildcards, with the correct alias for each command found replacing it's use in the syntax. If a command doesn't have an alias but is found with the wildcard then the extra notation isn't shown (see Split-Path example below)

```
PS>  gcm sp* -syn
sp (alias) -> Set-ItemProperty

sp [-Path] <string[]> [-Name] <string> [-Value] <Object> [-PassThru] [-Force] [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Credential <pscredential>] [-WhatIf] [-Confirm] [<CommonParameters>]

sp [-Path] <string[]> -InputObject <psobject> [-PassThru] [-Force] [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Credential <pscredential>] [-WhatIf] [-Confirm] [<CommonParameters>]

sp [-Name] <string> [-Value] <Object> -LiteralPath <string[]> [-PassThru] [-Force] [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Credential <pscredential>] [-WhatIf] [-Confirm] [<CommonParameters>]

sp -LiteralPath <string[]> -InputObject <psobject> [-PassThru] [-Force] [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] 
[-Credential <pscredential>] [-WhatIf] [-Confirm] [<CommonParameters>]


Split-Path [-Path] <string[]> [-Parent] [-Resolve] [-Credential <pscredential>] [<CommonParameters>]  

Split-Path [-Path] <string[]> [-Leaf] [-Resolve] [-Credential <pscredential>] [<CommonParameters>]  
```

There is one odd case that I'm not sure what to do about, if an alias and the command it's for both match a wildcard then what should the output look like. An example is `write` and `Write-Output`, currently with this implementation it'll produce output like this when you run `gcm wri* -syn`:

```
write (alias) -> Write-Output

write [-InputObject] <psobject> [-NoEnumerate] [<CommonParameters>]
```

But I don't think it should, it should either produce the old output of just the `Write-Output` or it should output both. There is duplicate detection in place to prevent outputting multiple of the same command so I'm thinking we probably want to just output the actual `Write-Output` syntax without the extra alias information.

ChrisLGardner commented:
I've discovered another issue this will have, if you do something like `gcm get-help, sls -syntax` then it'll replace all the mentions of sls with Get-Help:

```
get-help (alias) -> Select-String

get-help [-Pattern] <string[]> [-Path] <string[]> [-SimpleMatch] [-CaseSensitive] [-Quiet] [-List] [-NoEmphasis] [-Include <string[]>] 
[-Exclude <string[]>] [-NotMatch] [-AllMatches] [-Encoding <Encoding>] [-Context <int[]>] [<CommonParameters>]

get-help [-Pattern] <string[]> -InputObject <psobject> [-SimpleMatch] [-CaseSensitive] [-Quiet] [-List] [-NoEmphasis] [-Include <string[]>] [-Exclude <string[]>] [-NotMatch] [-AllMatches] [-Encoding <Encoding>] [-Context <int[]>] [<CommonParameters>]
```

So I need to figure out a way to handle that but I know what's causing it so that's a start.

ChrisLGardner commented:
I think I've fixed all the edge cases I can, the only one that's outstanding is if someone does `gcm del, sp* -syntax` then they'll get back the right output but instead of `del` being used for the alias in the `Remove-Item` output it'll actually be `ri` due to the order they are in the alias table. This only happens on commands with multiple aliases and when using a wildcard as well. I can't imagine people are often going to look for syntax on multiple aliases and a wildcard at the same time so I didn't put much effort into figuring out how to fix that.

This is now ready for a review since I refactored it into a new method.

vexx32 commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -569,6 +567,80 @@ private void OutputResultsHelper(IEnumerable<CommandInfo> results)
+        {
+            PSObject syntax = PSObject.AsPSObject(command.Syntax);
+
+            if (this.Name != null && !Array.Exists(this.Name, name => name.Equals(command.Name, StringComparison.InvariantCultureIgnoreCase)))
```
Perhaps it could be easier to just do a `foreach` over the `Name` values and work with each separately?

ChrisLGardner replied:
There's already a foreach loop in the `OutputResultsHelper` method which sends one found command at a time to the output, via this new method if the `Syntax` switch is provided. So I'm working with one at a time but need to check if the command I'm being passed was declared as the actual command to find or an alias or a wildcard, so I can then output things in the correct way.

vexx32 replied:
So what's this condition meant to be checking? Just not sure why we're using Array.Exists here & wondering if the condition can be simplified. 🙂

ChrisLGardner replied:
It's checking if the command name that's been passed in is one that was specified by a user, if not then we have to assume they specified an alias or a wildcard and do some extra formatting for those, if it is then just go with the default formatting. 

So if a user runs `Get-Command -Name del -Syntax` the code will find del and the command it resolves to as Remove-Item and attempt to return that, but as the user specified del we want to fiddle with the output a bit to make it clear that's an alias but still give the Remove-Item syntax.

vexx32 replied:
Could you use `Array.IndexOf(Name, command.Name) > -1`? I don't know for sure, but I have a feeling using the predicate might be a tad slower, and it isn't very clear in terms of the intent of the condition imo. 🙂

ChrisLGardner commented:
@SteveL-MSFT or @adityapatwardhan Can I get one of you to review this PR when you get a chance please?

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:
```diff
@@ -103,3 +103,79 @@ Describe "Get-Command Feature tests" -Tag Feature {
+            process {
+                "Processing ${Name}"
+            }
+'@ | out-File TestDrive:\Test-GcmSyntax.ps1
```
NIT: Personally, would prefer the herestring to be assigned to a variable and use `Set-Content` on a different line as I think it's more readable

```suggestion
'@ | Out-File -Path TestDrive:\Test-GcmSyntax.ps1
```

ChrisLGardner replied:
Fixed.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:
```diff
@@ -103,3 +103,79 @@ Describe "Get-Command Feature tests" -Tag Feature {
+'@ | out-File TestDrive:\Test-GcmSyntax.ps1
+
+            # Now set up an alias for that file
+            new-alias tgs TestDrive:\Test-GcmSyntax.ps1
```
```suggestion
            new-alias tgs TestDrive:\Test-GcmSyntax.ps1
```
```suggestion
            New-Alias -Name tgs -Value TestDrive:\Test-GcmSyntax.ps1
```

ChrisLGardner replied:
Fixed.

SteveL-MSFT commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -569,6 +567,80 @@ private void OutputResultsHelper(IEnumerable<CommandInfo> results)
+                            string.Format("{0}{1}", externalScript.Path, Environment.NewLine),
+                            Environment.NewLine,
+                            command.Syntax.Replace(command.Name, aliasName));
+                        syntax = PSObject.AsPSObject(replacedSyntax);
```
You repeat this line in each case, you can have it outside the switch statement

ChrisLGardner replied:
They are slightly different as the output will need to vary depending on what is found for the alias/wildcard. If the found command is a script then it's path needs to be added along with an extra new line as the syntax property of commands already includes a leading new line.

SteveL-MSFT replied:
I'm referring specifically to:

```csharp
syntax = PSObject.AsPSObject(replacedSyntax);
```

Unless my eyes are missing something isn't it exactly the same in each case statement before the break?

ChrisLGardner replied:
Good spot, I hadn't noticed that. I've fixed that now to only happen once after the switch.

SteveL-MSFT commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -569,6 +567,80 @@ private void OutputResultsHelper(IEnumerable<CommandInfo> results)
+                    }
+                    else if (_nameContainsWildcard && tableEntry.Value.Definition == command.Name)
+                    {
+                        aliasName = tableEntry.Key;
```
You have same result on line 593, you can just || this with line 590

ChrisLGardner replied:
Fixed.

SteveL-MSFT commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -569,6 +567,80 @@ private void OutputResultsHelper(IEnumerable<CommandInfo> results)
+        {
+            PSObject syntax = PSObject.AsPSObject(command.Syntax);
+
+            if (this.Name != null && !Array.Exists(this.Name, name => name.Equals(command.Name, StringComparison.InvariantCultureIgnoreCase)))
```
Should add a comment here explaining this line like you explained it in the resolved conversation

ChrisLGardner replied:
Added.

iSazonov commented:
@SteveL-MSFT We need a conclusion about the change. I have a strong concern. I am sure that Get-Command cmdlet is for script scenarios and should return typed output for aliases and syntax. For human scenarios there is Get-Help cmdlet.

ChrisLGardner commented:
Get-Command currently outputs typed objects except with -Syntax. I don't
see a time where you'd use that switch in a script and would be a breaking
change if it did return something other than a string as it does currently.

On Fri, Nov 1, 2019, 18:36 Ilya <notifications@github.com> wrote:

> @SteveL-MSFT <https://github.com/SteveL-MSFT> We need a conclusion about
> the change. I have a strong concern. I am sure that Get-Command cmdlet is
> for script scenarios and should return typed output for aliases and syntax.
> For human scenarios there is Get-Help cmdlet.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PowerShell/PowerShell/pull/10784?email_source=notifications&email_token=AEAOUFQSYY3CO7VFQR443WTQRRZKRA5CNFSM4JAIKSR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEC3ZP7A#issuecomment-548902908>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEAOUFQMHJVPDVVPCB56N5LQRRZKRANCNFSM4JAIKSRQ>
> .
>

SteveL-MSFT commented:
@iSazonov `Get-Command <x> -Syntax` returns string today as @ChrisLGardner says unless I'm misunderstanding your concern

iSazonov commented:
@SteveL-MSFT If we look other parameters Get-Command returns __typed__ output like:
```powershell
 (gcm -verb Update)[0].GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    FunctionInfo                             System.Management.Automation.CommandInfo
```
I am sure that this was and is a design intention for this cmdlet. Improvements for Syntax parameter must be in the direction too - return a typed object. To get user friendly output we could add new elements in formatting system for the typed object or improve Get-Help cmdlet.

vexx32 commented:
@iSazonov @SteveL-MSFT  there may well be a useful advantage in having the cmdlet return typed output.

Be that as it may, I don't think it is a good idea to do it in this PR. That work, if we want it, should be in another PR. To be particularly useful, I would imagine such an object would likely output objects that contain information about each cmdlet's name, parameter sets, and mandatory and optional parameters for each set, as well as position information (if relevant), type information, and possibly validation / transformation attributes for each parameter.

This kind of change should be its own work item and may warrant discussion in an issue to help determine what users would find useful. I don't think we will be gaining anything by attempting to do that work in this PR.

As it stands, this PR is an improvement, and we can build upon it further after this PR if we feel that typed output should be our goal.

ChrisLGardner commented:
> @iSazonov @SteveL-MSFT there may well be a useful advantage in having the cmdlet return typed output.
> 
> Be that as it may, I don't think it is a good idea to do it in this PR. That work, if we want it, should be in another PR. To be particularly useful, I would imagine such an object would likely output objects that contain information about each cmdlet's name, parameter sets, and mandatory and optional parameters for each set, as well as position information (if relevant), type information, and possibly validation / transformation attributes for each parameter.
> 
> This kind of change should be its own work item and may warrant discussion in an issue to help determine what users would find useful. I don't think we will be gaining anything by attempting to do that work in this PR.
> 
> As it stands, this PR is an improvement, and we can build upon it further after this PR if we feel that typed output should be our goal.

My issue with that is that it is still a breaking change. Get-Command already returns all of that information when you don't specify the `-Syntax` switch, even when specifying an alias or wildcard. If you're wanting to work with the syntax of a command in a programmatic way then you're going to ignore the `-Syntax` switch and work directly with the Parameters attribute (or write regex to parse the `-Syntax` output but that seems like a mistake given Parameters exists).

kilasuit commented:
> @SteveL-MSFT We need a conclusion about the change. I have a strong concern. I am sure that Get-Command cmdlet is for script scenarios and should return typed output for aliases and syntax. For human scenarios there is Get-Help cmdlet.

@iSazonov i think this concern and the other comments made as to what we should change and why are based on an incorrect assumption as to how people use Get-Command today not how they could use it in the future

This change makes sense for those that use Get-Command interactively with the Syntax parameter today and not for future changes.

> If we look other parameters Get-Command returns typed output

Whilst that's true there's no binding contract that states there must be similar outputs when differing parameters are provided to a command and therefore we should not look to change this behaviour

> Improvements for Syntax parameter must be in the direction too - return a typed object.

Disagree as genuinely there's no benefit within this PR to do so, not without breaking the current experience. We should get a change like this in and then look to iterate on new additional functionality around this area via additional switches for differing output types in future.

SteveL-MSFT commented:
@iSazonov I think we all can agree that an object should be returned, but is outside the scope of this PR since returning a string is the existing behavior

SteveL-MSFT reviewed (approved):
LGTM

vexx32 commented:
@PoshChan please retry windows

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-Windows`

adityapatwardhan commented:
@ChrisLGardner I tried building your branch and am seeing some issues. This happens when a new pwsh.exe is started and the first command is `Get-Command del -Syntax`. Please have a look:

```PowerShell
PS> Get-Command del -Syntax
Get-Command: The term 'del' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Suggestion [4,General]: The most similar commands are: del, dir, dbp, fl, gal, gl, iex, nal, sal, sl.
PS> Get-Command del -Syntax
Get-Command: The term 'del' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Suggestion [4,General]: The most similar commands are: del, dir, dbp, fl, gal, gl, iex, nal, sal, sl.
PS> Get-Command del -Syntax
Get-Command: The term 'del' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Suggestion [4,General]: The most similar commands are: del, dir, dbp, fl, gal, gl, iex, nal, sal, sl.
PS> Get-Command del -Syntax
Get-Command: The term 'del' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

Suggestion [4,General]: The most similar commands are: del, dir, dbp, fl, gal, gl, iex, nal, sal, sl.
PS> Get-Command del

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           del -> Remove-Item

PS> Get-Command del -Syntax
del (alias) -> Remove-Item

del [-Path] <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]

del -LiteralPath <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]
```

ChrisLGardner commented:
That's really weird, I'd not seen it before because it was auto loading my profile and that must be doing something to get around that. 

I'll have a look into it over the next day or two and see if I can figure out what's going on there.

ChrisLGardner commented:
@adityapatwardhan It looks like an issue in the session state and is in Preview 5 too. It looks like not all aliases are having their ResolvedCommand and other properties populated properly until after at least one command is successfully ran first. If you do `Get-Alias gci | Format-List *` you'll see ResolveCommand and a bunch of others are empty and that's what's causing this to fail.

I'll log a separate issue about this as it's outside the scope of this PR and way out of my knowledge area.

KirkMunro commented:
@ChrisLGardner I took a quick look at this, and the issue is module auto-loading.

In a new session loaded without a profile, if I invoke `gcm -syntax saps`, I get information about `saps` being an alias for `Start-Process`, but the `Microsoft.PowerShell.Management` module does not auto-load.

If in that same session I invoke `gcm -syntax Start-Process`, then I get the module is auto-loaded and I get the command syntax.

This makes me think it might be in scope for this PR. Whatever `Get-Command` does when you pass it a command name that results in the module auto-loading should happen when you pass it an alias name, with it auto-loading the module that contains that command before it pulls information about that command so that it can show the proper information.

Note that I say all of this without looking at the code, so this is theoretical, but I think that this work is related enough to this PR and shouldn't be too difficult to figure out since `gcm` already worked for cmdlets in modules that were not loaded yet.

ChrisLGardner committed:
Add check for null resolved name and lookup actual command info object if module is not imported

ChrisLGardner commented:
@KirkMunro thanks for the pointer there, I managed to find the method in CommandDiscovery that was already being used for command resolution and added an extra check in to call that if it hasn't resolved the alias to a CommandInfo.

@adityapatwardhan looks like I've resolved that bug if you get a chance to look at this again at some point.

adityapatwardhan commented on src/System.Management.Automation/engine/GetCommandCommand.cs:
```diff
@@ -1306,7 +1306,9 @@ private bool IsCommandMatch(ref CommandInfo current, out bool isDuplicate)
+                        current = ai.ResolvedCommand == null ?
+                            CommandDiscovery.LookupCommandInfo(ai.UnresolvedCommandName, this.MyInvocation.CommandOrigin, this.Context) :
+                            ai.ResolvedCommand;
                         if (current == null)
```
Seems this will be dead code now? Is that correct?

adityapatwardhan replied:
@ChrisLGardner ping..

ChrisLGardner replied:
No, doing some testing and it's there to resolve the command name if the module the alias is in hasn't been loaded yet and then that can be used to get the syntax. Without it you get the below behaviour:

```
PS> gcm del -syn
Remove-Item
```

with it you get this:

```
PS>  gcm del -syn
del (alias) -> Remove-Item

del [-Path] <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] 

del -LiteralPath <string[]> [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Recurse] [-Force] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Stream <string[]>] [<CommonParameters>]
```

vexx32 replied:
@ChrisLGardner I think @adityapatwardhan is referring to the line 1312 `if (current == null)` code path, which shouldn't be reached anymore since you're checking for null and accounting for that directly. If whatever's in that code path isn't reachable anymore, it should be removed. 🙂 

Also, this condition with the null can be simplified:

```csharp
current = ai.ResolvedCommand ?? CommandDiscovery.LookupCommandInfo(
    ai.UnresolvedCommandName,
    this.MyInvocation.CommandOrigin,
    this.Context);
```

ChrisLGardner replied:
Good point, I'll get that fixed now.

ChrisLGardner replied:
Looking at the failure in the tests there's a reason that check is there. It can sometimes return null from LookupCommandInfo. In the case for the tests (and locally) it is because `spvm` has a null `ResolvedCommand` and `UnresolvedCommandName`, probably in my case because I've apparently got both 1.1 and 2.0.0.0 of Hyper-V installed and I'd guess the AzDOS agents are similar.

vexx32 replied:
Might want to add a comment explaining that so nobody tries to remove that later 🙂

ChrisLGardner commented:
Looks like the last failing build is related to Test-Connection so not impact by the change I've made. though I'm certainly well behind master now but I can rebase if required.

vexx32 commented:
@PoshChan please retry macos

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-macOS`
